<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** CRISTIAN VAZQUEZ PLESENT & SOFIA HERNANDEZ ZATARAIN

**Fecha:** 16 de febrero del 2023.

**Expediente** : 731324 & 727151-.
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: https://github.com/sofiahzt/ProyectoConjunto_CVAZQUEZ_SHERNANDEZ.git
# Tarea 3: Clase 6

> ## Usando notebook de jupyter

> Usando compresión de listas o funciones map(sino recuerda como funciona observar el siguiente enlace https://www.pythonforbeginners.com/lists/list-comprehensions-in-python/), resolver los siguientes ejercicios:

>1. Resolver la siguiente ecuación recursiva usando funciones como se vió en clase
$$ D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2} \quad n\ge 3$$
con $D_1=0$ y $D_2 = 1$
>3. Count the number of spaces in the following string `variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores`.
>4. Remove all of the vowels in a string [make a list of the non-vowels].
>5. Find all of the words in a string that are less than 4 letters.
>6. Use a dictionary comprehension to count the length of each word in a sentence.
>7. Use a nested list comprehension to find all of the numbers from 1-1000 that are divisible by any single digit besides 1 (2-9). 

### Librerías

In [2]:
import numpy as np
from functools import reduce

### Ejercicio 1:

### Código de solución estudiante 1

>1. Resolver la siguiente ecuación recursiva usando funciones como se vió en clase
$$ D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2} \quad n\ge 3$$
con $D_1=0$ y $D_2 = 1$

In [3]:
# Código de solución CRISTIAN

def D(n):
    if n == 1:
        return 0
    elif n == 2:
        return 1
    else:
        return (n-1)*D(n-1) + (n-1)*D(n-2)

# Ejemplo de uso
print(D(5))  

44


>3. Count the number of spaces in the following string `variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores`.

In [4]:
# Código de solución CRISTIAN
frase = "variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores"
num_spacio = sum(map(lambda c: 1 if c == ' ' else 0, frase))
print(num_spacio)

26


>4. Remove all of the vowels in a string [make a list of the non-vowels].

In [5]:
# Código de solución CRISTIAN
frase = "Hola, ¿cómo estás?, esto es una prueba a ver si sirve"
vocales = "aeiouAEIOU"
consonante = [c for c in frase if c not in vocales]
print(consonante)

['H', 'l', ',', ' ', '¿', 'c', 'ó', 'm', ' ', 's', 't', 'á', 's', '?', ',', ' ', 's', 't', ' ', 's', ' ', 'n', ' ', 'p', 'r', 'b', ' ', ' ', 'v', 'r', ' ', 's', ' ', 's', 'r', 'v']


>5. Find all of the words in a string that are less than 4 letters.

In [6]:
# Código de solución CRISTIAN
string = "Esta es la frase que usaremos para ver si este codigo si funciona sino lloramos al final"
palabra_corta = [word for word in string.split() if len(word) < 4]
print(palabra_corta)

['es', 'la', 'que', 'ver', 'si', 'si', 'al']


>6. Use a dictionary comprehension to count the length of each word in a sentence.

In [7]:
# Código de solución CRISTIAN

frase = "Esta es la frase que usaremos para ver si este codigo si funciona sino lloramos al final"
contar = {word: len(word) for word in frase.split()}
print(contar)

{'Esta': 4, 'es': 2, 'la': 2, 'frase': 5, 'que': 3, 'usaremos': 8, 'para': 4, 'ver': 3, 'si': 2, 'este': 4, 'codigo': 6, 'funciona': 8, 'sino': 4, 'lloramos': 8, 'al': 2, 'final': 5}


>7. Use a nested list comprehension to find all of the numbers from 1-1000 that are divisible by any single digit besides 1 (2-9). 

In [8]:
# Código de solución CRISTIAN
numerod = [number for number in range(1, 1001) if any(number % digit == 0 for digit in range(2, 10))]
print(numerod)

[2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 15, 16, 18, 20, 21, 22, 24, 25, 26, 27, 28, 30, 32, 33, 34, 35, 36, 38, 39, 40, 42, 44, 45, 46, 48, 49, 50, 51, 52, 54, 55, 56, 57, 58, 60, 62, 63, 64, 65, 66, 68, 69, 70, 72, 74, 75, 76, 77, 78, 80, 81, 82, 84, 85, 86, 87, 88, 90, 91, 92, 93, 94, 95, 96, 98, 99, 100, 102, 104, 105, 106, 108, 110, 111, 112, 114, 115, 116, 117, 118, 119, 120, 122, 123, 124, 125, 126, 128, 129, 130, 132, 133, 134, 135, 136, 138, 140, 141, 142, 144, 145, 146, 147, 148, 150, 152, 153, 154, 155, 156, 158, 159, 160, 161, 162, 164, 165, 166, 168, 170, 171, 172, 174, 175, 176, 177, 178, 180, 182, 183, 184, 185, 186, 188, 189, 190, 192, 194, 195, 196, 198, 200, 201, 202, 203, 204, 205, 206, 207, 208, 210, 212, 213, 214, 215, 216, 217, 218, 219, 220, 222, 224, 225, 226, 228, 230, 231, 232, 234, 235, 236, 237, 238, 240, 242, 243, 244, 245, 246, 248, 249, 250, 252, 254, 255, 256, 258, 259, 260, 261, 262, 264, 265, 266, 267, 268, 270, 272, 273, 274, 275, 276, 278, 279, 280, 282,

### Código de solución estudiante 2

>1. Resolver la siguiente ecuación recursiva usando funciones como se vió en clase
$$ D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2} \quad n\ge 3$$
con $D_1=0$ y $D_2 = 1$

In [9]:
# Código se solución SOFÍA

def ecuacion_recursiva(n):
    d = np.zeros(n)
    d[0] = 0
    d[1] = 1
    
    def sol_eq(i):
        nonlocal d
        d[i-1] = (i-1) * d[i-2] + (i-1) * d[i-3]
        
    for i in range(3, n+1):
        sol_eq(i)
    
    return d

ecuacion_recursiva(6)

array([  0.,   1.,   2.,   9.,  44., 265.])

>3. Count the number of spaces in the following string `variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores`.

In [10]:
# Código de solución SOFÍA

string_spaces = "variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores"

string_spaces.count(" ")

26

>4. Remove all of the vowels in a string [make a list of the non-vowels].

In [11]:
# Código de solución SOFÍA

string_vowels = "Hola, ¿cómo estás?, esto es una prueba a ver si sirve"

sol_vowels = list([cons for cons in string_vowels if cons not in
            ['a','e', 'i', 'o', 'u','A', 'E', 'I', 'O', 'U', 'á', 'é', 'í', 'ó', 'ú', 'Á', 'É', 'Í', 'Ó', 'Ú', ' ']])

sol_vowels

['H',
 'l',
 ',',
 '¿',
 'c',
 'm',
 's',
 't',
 's',
 '?',
 ',',
 's',
 't',
 's',
 'n',
 'p',
 'r',
 'b',
 'v',
 'r',
 's',
 's',
 'r',
 'v']

>5. Find all of the words in a string that are less than 4 letters.

In [12]:
# Código de solución SOFÍA

string_len = "Esta es la frase que usaremos para ver si este codigo si funciona, si no, lloramos al final"

sol_len = list(filter(lambda palabra: len(palabra) < 4, string_len.split()))

sol_len

['es', 'la', 'que', 'ver', 'si', 'si', 'si', 'no,', 'al']

>6. Use a dictionary comprehension to count the length of each word in a sentence.

In [13]:
# Código de solución SOFÍA

string_count = "Esta es la frase que usaremos para ver si este codigo si funciona, si no, lloramos al final"

diccionario = {palabra: len(palabra) for palabra in string_count.split()}

diccionario

{'Esta': 4,
 'es': 2,
 'la': 2,
 'frase': 5,
 'que': 3,
 'usaremos': 8,
 'para': 4,
 'ver': 3,
 'si': 2,
 'este': 4,
 'codigo': 6,
 'funciona,': 9,
 'no,': 3,
 'lloramos': 8,
 'al': 2,
 'final': 5}

>7. Use a nested list comprehension to find all of the numbers from 1-1000 that are divisible by any single digit besides 1 (2-9). 

In [14]:
# Código de solución SOFÍA

lista = [n for n in range(1,1001) if [True for i in range(2,10) if n % i == 0]]

lista

[2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 12,
 14,
 15,
 16,
 18,
 20,
 21,
 22,
 24,
 25,
 26,
 27,
 28,
 30,
 32,
 33,
 34,
 35,
 36,
 38,
 39,
 40,
 42,
 44,
 45,
 46,
 48,
 49,
 50,
 51,
 52,
 54,
 55,
 56,
 57,
 58,
 60,
 62,
 63,
 64,
 65,
 66,
 68,
 69,
 70,
 72,
 74,
 75,
 76,
 77,
 78,
 80,
 81,
 82,
 84,
 85,
 86,
 87,
 88,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 98,
 99,
 100,
 102,
 104,
 105,
 106,
 108,
 110,
 111,
 112,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 122,
 123,
 124,
 125,
 126,
 128,
 129,
 130,
 132,
 133,
 134,
 135,
 136,
 138,
 140,
 141,
 142,
 144,
 145,
 146,
 147,
 148,
 150,
 152,
 153,
 154,
 155,
 156,
 158,
 159,
 160,
 161,
 162,
 164,
 165,
 166,
 168,
 170,
 171,
 172,
 174,
 175,
 176,
 177,
 178,
 180,
 182,
 183,
 184,
 185,
 186,
 188,
 189,
 190,
 192,
 194,
 195,
 196,
 198,
 200,
 201,
 202,
 203,
 204,
 205,
 206,
 207,
 208,
 210,
 212,
 213,
 214,
 215,
 216,
 217,
 218,
 219,
 220,
 222,
 224,
 225,
 226,
 228,
 230,
 231,
 232,
 234,
 235,

<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** CRISTIAN VAZQUEZ PLESENT & SOFIA HERNANDEZ ZATARAIN

**Fecha:** 16 de febrero del 2023.

**Expediente** : 731324 & 727151-.
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: Link con el enlace del repositorio del creador del proyecto

# Tarea 4: Clase 7



# 1
Como ejemplo simple de una simulación de Monte Carlo, considere calcular la probabilidad de una suma particular del lanzamiento de tres dados (cada dado tiene valores del uno al seis). Además cada dado tiene las siguientes carácterísticas: el primer dado no está cargado (distribución uniforme todos son equiprobables); el segundo y tercer dado están cargados basados en una distribución binomial con parámetros (`n=5, p=0.5` y `n=5, p=0.2`). Calcule la probabilidad de que la suma resultante sea 7, 14 o 18. 


# 2 Ejercicio de aplicación- Cafetería Central

Premisas para la simulación:
- Negocio de alimentos que vende bebidas y alimentos.
- Negocio dentro del ITESO.
- Negocio en cafetería central.
- Tipo de clientes (hombres y mujeres).
- Rentabilidad del 60%.

## Objetivo
Realizar una simulación estimado el tiempo medio que se tardaran los clientes en ser atendidos entre el horario de 6:30 a 1 pm. Además saber el consumo. 
**Analizar supuestos y limitantes**

## Supuestos en simulación
Clasificación de clientes: 
- Mujer  = 1 $\longrightarrow$ aleatorio < 0.5
- Hombre = 0 $\longrightarrow$ aleatorio $\geq$ 0.5.

Condiciones iniciales:
- Todas las distrubuciones de probabilidad se supondrán uniformes.
- Tiempo de simulación: 6:30 am - 1:30pm $\longrightarrow$ T = 7 horas = 25200 seg.
- Tiempo de llegada hasta ser atendido: Min=5seg, Max=30seg.
- Tiempo que tardan los clientes en ser atendidos:
    - Mujer: Min = 1 min= 60seg, Max = 5 min = 300 seg
    - Hombre: Min = 40 seg, Max = 2 min= 120 seg
- Consumo según el tipo de cliente:
    - Mujer: Min = 30 pesos, Max = 100 pesos
    - Hombre: Min = 20 pesos, Max = 80 pesos

Responder las siguientes preguntas basados en los datos del problema:
1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.
2. ¿Cuál fue el consumo promedio de los hombres y mujeres?
3. ¿Cuál fue el número de personas atendidas por día?
4. ¿Cuál fue el tiempo de atención promedio?
5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?

### Ejercicio 1: 

>Como ejemplo simple de una simulación de Monte Carlo, considere calcular la probabilidad de una suma particular del lanzamiento de tres dados (cada dado tiene valores del uno al seis). Además cada dado tiene las siguientes carácterísticas: el primer dado no está cargado (distribución uniforme todos son equiprobables); el segundo y tercer dado están cargados basados en una distribución binomial con parámetros (`n=5, p=0.5` y `n=5, p=0.2`). Calcule la probabilidad de que la suma resultante sea 7, 14 o 18. 

### Código de solución estudiante 1

In [15]:
# Código de solución CRISTIAN

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
import scipy.stats as st

In [16]:
# Código de solución CRISTIAN

D1= np.random.randint(1,7,100000)
DC1= st.binom(n=5,p=0.5, loc = 1).rvs(size=100000)
DC2= st.binom(n=5,p=0.2, loc = 1).rvs(size=100000)

Sum= D1 + DC1 + DC2

prob1 = sum(list(map(lambda x: 1 if x == 7 else 0, Sum))) / 100000
prob2 = sum(list(map(lambda x: 1 if x == 14 else 0, Sum))) / 100000
prob3 = sum(list(map(lambda x: 1 if x == 18 else 0, Sum))) / 100000

np.array([prob1,prob2,prob3])

array([0.12648, 0.01363, 0.     ])

#### Interpretacion 

- La probabilidad que sea 7 es de 12.72%
- La probabilidad que sea resultante sea 14 es de 1.36%
- La probabilidad que sea resultante sea 18 es de 0%


### Código de solución estudiante 2

In [17]:
# Código de solución estudiante 2


In [18]:
# Código de solución SOFIA

n = 100000

dado1 = np.random.randint(1, 7, size = n)
dado2 = np.random.binomial(5, 0.5, size = n)
dado3 = np.random.binomial(5, 0.2, size = n)

suma = dado1 + dado2 + dado3

c7 = np.count_nonzero(suma == 7)
c14 = np.count_nonzero(suma == 14)
c18 = np.count_nonzero(suma == 18)

prob7 = c7 / n
prob14 = c14 / n
prob18 = c18 / n

prob7, prob14, prob18

(0.16147, 0.00046, 0.0)

#### Interpretacion 

- La probabilidad que sea 7 es de: $ 16.10% $
- La probabilidad que sea resultante sea 14 es de: $ 0.062% $
- La probabilidad que sea resultante sea 18 es de: $ 0% $

### Ejercicio 2: de aplicación- Cafetería Central

Premisas para la simulación:
- Negocio de alimentos que vende bebidas y alimentos.
- Negocio dentro del ITESO.
- Negocio en cafetería central.
- Tipo de clientes (hombres y mujeres).
- Rentabilidad del 60%.

### Objetivo
Realizar una simulación estimado el tiempo medio que se tardaran los clientes en ser atendidos entre el horario de 6:30 a 1 pm. Además saber el consumo. 
**Analizar supuestos y limitantes**

### Supuestos en simulación
Clasificación de clientes: 
- Mujer  = 1 $\longrightarrow$ aleatorio < 0.5
- Hombre = 0 $\longrightarrow$ aleatorio $\geq$ 0.5.

Condiciones iniciales:
- Todas las distrubuciones de probabilidad se supondrán uniformes.
- Tiempo de simulación: 6:30 am - 1:30pm $\longrightarrow$ T = 7 horas = 25200 seg.
- Tiempo de llegada hasta ser atendido: Min=5seg, Max=30seg.
- Tiempo que tardan los clientes en ser atendidos:
    - Mujer: Min = 1 min= 60seg, Max = 5 min = 300 seg
    - Hombre: Min = 40 seg, Max = 2 min= 120 seg
- Consumo según el tipo de cliente:
    - Mujer: Min = 30 pesos, Max = 100 pesos
    - Hombre: Min = 20 pesos, Max = 80 pesos

Responder las siguientes preguntas basados en los datos del problema:
1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.
2. ¿Cuál fue el consumo promedio de los hombres y mujeres?
3. ¿Cuál fue el número de personas atendidas por día?
4. ¿Cuál fue el tiempo de atención promedio?
5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?


### Código de solución estudiante 1

In [3]:
# Código de solución CRISTIAN
# Datos del problema
d = 5
T =25200
T_at_min = 5; T_at_max = 30
T_mujer_min =60; T_mujer_max = 300
T_hombre_min = 40; T_hombre_max = 120
C_mujer_min = 30; C_mujer_max = 100
C_hombre_min = 20; C_hombre_max = 80


In [7]:
# Código de solución CRISTIAN
def tiempo():
    t = np.random.randint(T_at_min,T_at_max+1)
    customer = np.random.random()
    if 0<=customer<0.5:
        mt= np.random.randint(T_mujer_min,T_mujer_max+1)
        tiempo = t + mt 
    if 0.5<=customer:
        ht= np.random.randint(T_hombre_min,T_hombre_max+1)
        tiempo = t + ht
    return tiempo

def consumo(): 
    t = np.random.randint(T_at_min,T_at_max+1)
    customer = np.random.random()
    if 0<=customer<0.5:
        c= np.random.randint(C_mujer_min,C_mujer_max+1)
    if 0.5<=customer:
        c= np.random.randint(C_hombre_min,C_hombre_max+1)
    return c

In [8]:
# Código de solución CRISTIAN
#1 dia 
segundos = []
ventas=[]
personas=0
while sum(segundos)<T:
    segundos.append(tiempo())
    ventas.append(consumo())
    personas +=1

#5 dias 
segundos5 = []
ventas5=[]
personas5=0
while sum(segundos5)<(T*5):
    segundos5.append(tiempo())
    ventas5.append(consumo())
    personas5 +=1

### Responder las siguientes preguntas basados en los datos del problema:



> 1.- ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.


In [14]:
# Código de solución CRISTIAN

print('Los gastos de hombres y mujeres en 5 días de trabajo fue de $',sum(ventas5))

Los gastos de hombres y mujeres en 5 días de trabajo fue de $ 48353


> 2. ¿Cuál fue el consumo promedio de los hombres y mujeres?

In [15]:
# Código de solución CRISTIAN

print('El consumo promedio de  hombres y mujeres fue de$',sum(ventas)/personas)

El consumo promedio de  hombres y mujeres fue de$ 56.06748466257669


> 3. ¿Cuál fue el número de personas atendidas por día?

In [16]:
# Código de solución CRISTIAN

print('El numero de personas atendidas fue', personas)

El numero de personas atendidas fue 163


> 4. ¿Cuál fue el tiempo de atención promedio?

In [17]:
# Código de solución CRISTIAN

print('El  tiempo promedio de atencion fue', sum(segundos)/personas,'segundos')

El  tiempo promedio de atencion fue 155.74846625766872 segundos


> 5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?

In [18]:
# Código de solución CRISTIAN

print('La ganancia promedio de la cafetería fue $',sum(ventas5)/5,
     'La rentabilidad fue de $',sum(ventas5)/5*.60)

La ganancia promedio de la cafetería fue $ 9670.6 La rentabilidad fue de $ 5802.36


### Código de solución estudiante 2

In [19]:
d = 5
T =25200
T_at_min = 5; T_at_max = 30
T_mujer_min =60; T_mujer_max = 300
T_hombre_min = 40; T_hombre_max = 120
C_mujer_min = 30; C_mujer_max = 100
C_hombre_min = 20; C_hombre_max = 80

In [20]:
def simulaciont():
           
        T = np.random.uniform(T_at_min, T_at_max + 1)

        mujer = 1 if np.random.uniform() < 0.5 else 0

        if mujer:
            tiempo = np.random.uniform(T_mujer_min, T_mujer_max + 1)
            mujeres_atendidas = T + tiempo
            
        else:
            tiempo = np.random.uniform(T_hombre_min, T_hombre_max + 1)
            hombres_atendidos = T + tiempo
        
        return simulaciont 

def simulacionc():
    
    T = np.random.uniform(T_at_min, T_at_max + 1)
    
    mujer = 1 if np.random.uniform() < 0.5 else 0

    if mujer:
            consumo = np.random.uniform(C_mujer_min, C_mujer_max + 1)
            
    else:
            tiempo = np.random.uniform(T_hombre_min, T_hombre_max + 1)
        
    return simulacionc

### Responder las siguientes preguntas basados en los datos del problema:


> 1.- ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.

> 2. ¿Cuál fue el consumo promedio de los hombres y mujeres?

> 3. ¿Cuál fue el número de personas atendidas por día?

> 4. ¿Cuál fue el tiempo de atención promedio?

> 5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?